# Rapport de projet – Analyse de données et apprentissage automatique

## Objectif général

Ce projet a pour objectif de conduire un pipeline complet d’analyse de données, depuis le choix du dataset jusqu’à l’application d’un ou plusieurs modèles de machine learning.Vous travaillerez en groupe (3 à 4 étudiants) et documenterez toutes les étapes de votre démarche dans ce rapport.Chaque section doit être remplie de manière rigoureuse, avec des explications claires et des justifications.Le rapport servira à l’évaluation finale du module pour **40 % de la note finale**.

# Partie 1 – Définition du sujet et choix du dataset

## 1.1 Thématique générale

**Domaine choisi :**  
Le projet s’inscrit dans le domaine de l’économie et de la finance, plus précisément dans l’analyse du risque de crédit.

**Problématique :**  
L’objectif de ce projet est d’analyser les caractéristiques financières et personnelles des emprunteurs afin de prédire le risque de défaut de paiement.  
La question centrale étudiée est la suivante :  
**Quels facteurs influencent le défaut de paiement et dans quelle mesure peut-on anticiper ce risque à partir des données disponibles ?**

**Intérêt du sujet :**  
L’évaluation du risque de crédit constitue un enjeu majeur pour les établissements financiers. Une mauvaise estimation du risque peut entraîner des pertes économiques importantes, tandis qu’une estimation trop stricte peut conduire à refuser l’accès au crédit à des emprunteurs pourtant solvables.

**Intérêt pour la société et les entreprises :**  
Ce sujet présente un intérêt à la fois économique et sociétal. Il permet d’améliorer la stabilité financière des institutions, de favoriser des décisions d’octroi de crédit plus justes et cohérentes, de limiter le surendettement des ménages et d’optimiser les politiques de gestion du risque.

## 1.2 Recherche et sélection du dataset

### Informations générales sur le dataset

**Nom du dataset :** Credit Risk Dataset  

**Source et lien d’accès :** Kaggle  

**Auteur / organisation :** Communauté Kaggle  

**Taille :** 32 581 lignes et 12 colonnes  

**Format du fichier :** CSV  

### Vérification de la qualité

Le dataset contient un volume de données suffisant pour réaliser une analyse statistique approfondie et entraîner des modèles de machine learning.  
Les variables sont clairement nommées et compréhensibles.  
Il comporte une variable cible binaire (`loan_status`) permettant d’identifier les situations de défaut de paiement.  
Certaines valeurs manquantes sont présentes, mais leur proportion reste limitée et traitable.

### Justification du choix

Ce dataset a été choisi car il est directement adapté à la problématique du risque de crédit. Il inclut des variables financières importantes (revenu, montant du prêt, taux d’intérêt), des informations sur les emprunteurs (âge, ancienneté d’emploi, historique de crédit) ainsi qu’une variable cible liée au défaut de paiement.

**Limites ou difficultés potentielles :** 

Le dataset présente des valeurs manquantes sur certaines variables, notamment l’ancienneté d’emploi et le taux d’intérêt.  
De plus, les données étant issues d’une source unique et anonymisées, des biais d’échantillonnage peuvent exister.

## 1.3 Validation du dataset

| Critère        | Question                                                                 | Réponse (Oui/Non) | Détail / Justification |
|---------------|--------------------------------------------------------------------------|-------------------|------------------------|
| Pertinence    | Le dataset permet-il de répondre à votre question de départ ?            | Oui               | Présence d’une variable cible liée au défaut de paiement |
| Clarté        | Les variables sont-elles bien nommées et compréhensibles ?               | Oui               | Noms explicites et cohérents avec le domaine du crédit |
| Propreté      | Les données semblent-elles utilisables sans nettoyage majeur ?           | Oui               | Valeurs manquantes présentes mais en proportion modérée |
| Taille        | Le dataset est-il d’une taille adaptée à votre analyse ?                 | Oui               | 32 581 observations, suffisant pour l’EDA et la modélisation |
| Accessibilité | Le format est-il compatible avec Python (CSV, XLSX) ?                    | Oui               | Format CSV directement lisible avec pandas |
| Actualité     | Les données sont-elles récentes ou encore valides ?                      | Oui               | Données réalistes et pertinentes pour une étude du risque |

# Partie 2 – Exploration initiale des données

## 2.1 Chargement et aperçu du dataset

In [1]:
import pandas as pd

PATH = 'credit_risk_dataset.csv'

df = pd.read_csv(PATH)

df.head()

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
0,22,59000,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3
1,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2
2,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3
3,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2
4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4


**Dimensions du dataset :**

On affiche le nombre de lignes (observations) et de colonnes (variables).

In [2]:
df.shape

(32581, 12)

**Variables :** nom et type

On liste les variables et leurs types pandas (numérique / catégorielle).

In [5]:
df.dtypes

person_age                      int64
person_income                   int64
person_home_ownership          object
person_emp_length             float64
loan_intent                    object
loan_grade                     object
loan_amnt                       int64
loan_int_rate                 float64
loan_status                     int64
loan_percent_income           float64
cb_person_default_on_file      object
cb_person_cred_hist_length      int64
dtype: object

**Signification de chaque variable :**

- **`person_age`** : âge du client (années)
- **`person_income`** : revenu annuel du client
- **`person_home_ownership`** : statut de logement (RENT / OWN / MORTGAGE / OTHER)
- **`person_emp_length`** : ancienneté professionnelle (années)
- **`loan_intent`** : objectif du prêt (EDUCATION, MEDICAL, PERSONAL, etc.)
- **`loan_grade`** : note de crédit du prêt (A → G, ordre de risque croissant)
- **`loan_amnt`** : montant du prêt
- **`loan_int_rate`** : taux d’intérêt du prêt
- **`loan_status`** : statut du prêt (cible) — **0 = non défaut, 1 = défaut**
- **`loan_percent_income`** : part du revenu consacrée au prêt (ratio)
- **`cb_person_default_on_file`** : défaut antérieur enregistré (Y / N)
- **`cb_person_cred_hist_length`** : durée de l’historique de crédit (années)

**Valeurs manquantes et incohérences :**

On calcule le nombre de valeurs manquantes par variable, puis le pourcentage par rapport au total.

In [6]:
missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)

missing_report = pd.DataFrame({
    'missing_count': missing,
    'missing_%': missing_pct
}).sort_values('missing_count', ascending=False)

missing_report

,missing_count,missing_%
loan_int_rate,3116,9.56
person_emp_length,895,2.75
person_age,0,0.00
person_income,0,0.00
person_home_ownership,0,0.00
loan_intent,0,0.00
loan_grade,0,0.00
loan_amnt,0,0.00
loan_status,0,0.00
loan_percent_income,0,0.00


**Résultat attendu :**
- `person_emp_length` : **895** valeurs manquantes (~**2,75 %**)
- `loan_int_rate` : **3 116** valeurs manquantes (~**9,56 %**)
- Toutes les autres variables : **0** valeur manquante

**Contrôles simples d’incohérences :**
On vérifie des contraintes de bon sens (valeurs négatives, âges irréalistes, etc.).

In [8]:
# Contrôles de cohérence
checks = {
    'age_negative': (df['person_age'] < 0).sum(),
    'income_negative': (df['person_income'] < 0).sum(),
    'loan_amnt_negative': (df['loan_amnt'] < 0).sum(),
    'int_rate_negative': (df['loan_int_rate'] < 0).sum(),
    'emp_length_negative': (df['person_emp_length'] < 0).sum(),
    'loan_percent_income_negative': (df['loan_percent_income'] < 0).sum(),
    'cred_hist_negative': (df['cb_person_cred_hist_length'] < 0).sum(),
}

pd.Series(checks)

age_negative                    0
income_negative                 0
loan_amnt_negative              0
int_rate_negative               0
emp_length_negative             0
loan_percent_income_negative    0
cred_hist_negative              0
dtype: int64

**Interprétation :** si ces compteurs sont à 0, il n’y a pas d’incohérence évidente sur ces règles.
Les valeurs extrêmes (outliers) peuvent exister sans être des erreurs : elles seront traitées lors de l’analyse exploratoire.

## 2.2 Typologie des données

Classement des variables selon leur type (quantitative continue/discrète, qualitative nominale/ordinale, etc.).

#### Variables quantitatives continues
- `person_income`
- `person_emp_length`
- `loan_amnt`
- `loan_int_rate`
- `loan_percent_income`

#### Variables quantitatives discrètes
- `person_age`
- `cb_person_cred_hist_length`
- `loan_status` (binaire 0/1)

#### Variables qualitatives nominales
- `person_home_ownership`
- `loan_intent`
- `cb_person_default_on_file`

#### Variables qualitatives ordinales
- `loan_grade` (A → G)

#### Variables temporelles
- Aucune variable de type date/année n’est présente dans ce dataset.

#### Autres
- Aucune donnée non structurée (texte libre, image, etc.).


### Variables les plus importantes pour l’analyse (justification)
Dans une analyse de risque de crédit, les variables les plus importantes sont généralement celles qui reflètent :
1) la **capacité de remboursement**
2) le **profil de risque du prêt**
3) l’**historique de crédit**
4) la **stabilité**

#### 1) Capacité de remboursement
- `person_income`
- `loan_percent_income`
- `loan_amnt`

#### 2) Risque du prêt
- `loan_int_rate`
- `loan_grade`
- `loan_intent`

#### 3) Historique de crédit
- `cb_person_default_on_file`
- `cb_person_cred_hist_length`

#### 4) Stabilité / profil socio-démographique
- `person_emp_length`
- `person_home_ownership`
- `person_age`

on peut afficher des statistiques descriptives rapides :

In [10]:
df.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
person_age,32581.0,NaN,NaN,NaN,27.7346,6.348078,20.0,23.0,26.0,30.0,144.0
person_income,32581.0,NaN,NaN,NaN,66074.84847,61983.119168,4000.0,38500.0,55000.0,79200.0,6000000.0
person_home_ownership,32581,4,RENT,16446,NaN,NaN,NaN,NaN,NaN,NaN,NaN
person_emp_length,31686.0,NaN,NaN,NaN,4.789686,4.14263,0.0,2.0,4.0,7.0,123.0
loan_intent,32581,6,EDUCATION,6453,NaN,NaN,NaN,NaN,NaN,NaN,NaN
loan_grade,32581,7,A,10777,NaN,NaN,NaN,NaN,NaN,NaN,NaN
loan_amnt,32581.0,NaN,NaN,NaN,9589.371106,6322.086646,500.0,5000.0,8000.0,12200.0,35000.0
loan_int_rate,29465.0,NaN,NaN,NaN,11.011695,3.240459,5.42,7.9,10.99,13.47,23.22
loan_status,32581.0,NaN,NaN,NaN,0.218164,0.413006,0.0,0.0,0.0,0.0,1.0
loan_percent_income,32581.0,NaN,NaN,NaN,0.170203,0.106782,0.0,0.09,0.15,0.23,0.83


### Variable cible
La variable cible (à prédire / expliquer) est **`loan_status`**.
- `0` : non défaut
- `1` : défaut

On peut vérifier la distribution de la cible :

In [11]:
df['loan_status'].value_counts(dropna=False).to_frame('count')

# En pourcentage
(df['loan_status'].value_counts(normalize=True) * 100).round(2).to_frame('pct')

,pct
loan_status,
0,78.18
1,21.82


# Partie 3 – Nettoyage et préparation du dataset

## 3.1 Gestion des valeurs manquantes

### Valeurs manquantes : nombre + pourcentage
missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)

missing_report = pd.DataFrame({
    'missing_count': missing,
    'missing_%': missing_pct
}).sort_values('missing_count', ascending=False)

missing_report

**Interprétation :**
Les valeurs manquantes concernent uniquement deux variables :

- `person_emp_length` : 895 valeurs manquantes (~2,75 %)
- `loan_int_rate` : 3 116 valeurs manquantes (~9,56 %)

Toutes les autres variables ne présentent aucune valeur manquante.

**Stratégie appliquée :**  
Imputation par la médiane pour les variables numériques.

**Justification :**  
La proportion de valeurs manquantes est modérée ; supprimer les lignes ferait perdre de l’information.  
La médiane est robuste face aux valeurs extrêmes, fréquentes dans les données financières.

In [12]:
# Imputation par la médiane
df_clean = df.copy()

df_clean["person_emp_length"] = df_clean["person_emp_length"].fillna(
    df_clean["person_emp_length"].median()
)

df_clean["loan_int_rate"] = df_clean["loan_int_rate"].fillna(
    df_clean["loan_int_rate"].median()
)

# Vérification
df_clean.isnull().sum()

person_age                    0
person_income                 0
person_home_ownership         0
person_emp_length             0
loan_intent                   0
loan_grade                    0
loan_amnt                     0
loan_int_rate                 0
loan_status                   0
loan_percent_income           0
cb_person_default_on_file     0
cb_person_cred_hist_length    0
dtype: int64

## 3.2 Détection et traitement des doublons

In [13]:
# Détection des doublons
df_clean.duplicated().sum()

165

**Interprétation :**
165 doublons ont été détectés avant traitement.

In [15]:
# Suppression des doublons
df_clean = df_clean.drop_duplicates()

df_clean.shape

(32416, 12)

Ils ont été supprimés afin d’éviter une sur-représentation artificielle de certaines observations, pouvant biaiser les statistiques et les modèles.

## 3.3 Détection des valeurs aberrantes

In [16]:
import numpy as np

# Colonnes numériques (incluant la cible pour l'instant)
numeric_cols = df_clean.select_dtypes(include=["int64", "float64"]).columns

# Détection via IQR
outlier_summary = []

for col in numeric_cols:
    q1 = df_clean[col].quantile(0.25)
    q3 = df_clean[col].quantile(0.75)
    iqr = q3 - q1
    
    low = q1 - 1.5 * iqr
    high = q3 + 1.5 * iqr
    
    outliers = ((df_clean[col] < low) | (df_clean[col] > high)).sum()
    outlier_summary.append((col, int(outliers), low, high))

outlier_df = pd.DataFrame(
    outlier_summary,
    columns=["variable", "n_outliers", "low_bound", "high_bound"]
).sort_values("n_outliers", ascending=False)

outlier_df


,variable,n_outliers,low_bound,high_bound
5,loan_status,7089,0.00,0.00
3,loan_amnt,1679,-5875.00,23125.00
0,person_age,1491,12.50,40.50
1,person_income,1478,-22472.00,140232.00
7,cb_person_cred_hist_length,1139,-4.50,15.50
2,person_emp_length,852,-5.50,14.50
6,loan_percent_income,650,-0.12,0.44
4,loan_int_rate,70,1.56,20.04


## 3.4 Encodage et mise à l’échelle des variables

Certaines variables doivent être converties en numériques avant d’être utilisées dans un modèle.Questions :Quelles colonnes ont été encodées ?Quelle méthode avez-vous utilisée ?Pourquoi est-il important de normaliser ou standardiser les données avant l’entraînement des modèles ?**Réponse (à rédiger et JUSTIFIER avec des extraits ou résultats du code) :**

# Partie 4 – Analyse exploratoire et visualisations

Réalisez une première exploration visuelle pour mieux comprendre les relations entre les variables.Questions :Quelles relations ou corrélations observez-vous ?Quelles variables semblent influencer la variable cible ?Quelles hypothèses pouvez-vous formuler ?Quelles tendances ou patterns remarquez-vous dans les graphiques ?**Réponse (à rédiger et JUSTIFIER avec des extraits ou résultats du code) :**